### PR TITLE
Replace httpproxy= url protocol option with CProxy class

### PR DESCRIPTION
**Pull request by jhol**
This pull request supercedes an earlier attempt (#8680) to fix issues with per-stream proxy servers when used with HLS streams.

Previous discussion (#8434) revealed that the `httpproxy` url parameter is a potential security hazard. If an attacker can insert a malicious URL into Kodi, for example via a malicious playlist served over the internet, the attacker can command the application to make a connection to a malicious proxy, potentially leaking sensitive information.

Therefore, this patch set removes the `httpproxy` url option altogether, replaces it with a new out-of-band method of handling proxy servers: the `CProxy` class.

The `CProxy` class handles parsing, formatting and storing proxy server descriptions, and supports the `IArchivable` and `ISerializable` interfaces.

Storing the server description in a single object reduces boiler-plate, as seen in `CCurlFile`, which has the `m_proxy`, `m_proxyuserpass` and `m_proxytype` member variables and accessor methods. These are replaced by a single member variable, and a single set of accessor methods.

A `CProxy` member variable has been added to `CURL`, because now that the proxy server description is carried outside the url, if `CProxy` were not added to `CURL`, the two objects would typically be handled as a pair throughout the code-base in method arguments and member variables.

Support for HTTP proxies has been added to `CDVDDemuxFFmpeg` via the `http_proxy` option which has been available since v3.0.

Finally, a new `proxy` property has been added to `XBMCAddon::CListItem`, so that add-ons may set the proxy through the Python `getProxy`/`setProxy` interface.

In future, this work can be extended by adding per-stream proxy support to more of the DVDDemuxers, and by extending support for the SOCKS protocols.

### CODE REVIEW

**Comment by FernetMenta** [comment]
Looks much too complicated. Why don't you use key/value pairs of list items?

**Comment by jhol** [comment]
@FernetMenta those key-value pairs need would need some kind of container, because they need to be handed around between different classes that handle protocol options in different ways. It also helps to have functions for parsing/formatting the key-value pairs from/to urls.

A class suits these requirements nicely.

The result is a reduction in proxy handling boiler-plate spread around the code-base, with the code mostly encapsulated inside `CProxy`.

File: xbmc/utils/Proxy.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+#include "Proxy.h"
+
+using std::string;
+
+const uint16_t DefaultPorts[] = {
+  3128,
+  1080,
+  1080,
+  1080,
+  1080
```
Comment (afedchin) [inline xbmc/utils/Proxy.cpp]: fourfold 1080?

**Comment by jhol** [comment]
@basrieter , @fritsch - any comments?

File: xbmc/utils/Proxy.cpp (hunk 1 above)
Comment (jhol) [inline xbmc/utils/Proxy.cpp]: @afedchin - yes all the versions SOCKS protocol use the same port.

**Comment by FernetMenta** [comment]
@jhol there won't be an extra proxy class. The new InputStream addon uses key/value pairs and I don't see the need for introducing an additional method just for the proxy.

**Comment by basrieter** [comment]
I think the idea looks good and it seems much better to do it on a per request bases. But I do have to agree with @FernetMenta that if there has been a decision on how to handle similar properties, this change should stick to that. 

@FernetMenta where can I read up on this new InputStream addon?

**Comment by FernetMenta** [comment]
@basrieter check out this: https://github.com/xbmc/xbmc/pull/9173

The addon gets those key/value pairs that it has registered for.
https://github.com/FernetMenta/xbmc/blob/inputstream/xbmc/addons/InputStream.cpp#L112

**Comment by basrieter** [comment]
So, from a Python add-on view, how would these two things work together? Or is the InputStream extension point just an add-on that tell Kodi how to handle a certain stream type and from a video add-on you just pass the stream and its type? 

...perhaps I should move this to https://github.com/xbmc/xbmc/pull/9173? I already have some streams that use MPD, so I could do some testing?

**Comment by FernetMenta** [comment]
@basrieter the InputStream Addon is an example how to use an already existing mechanism. i.e. set proxy.url, proxy.user, etc on the file item.

**Comment by basrieter** [comment]
@FernetMenta I am kinda lost here. I checked, but the kodi.inputstream Addon is a binary addon? But if I read it correctly, the ListItem.setProperty can now be used to pass proxy info?

**Pull request closed without merging**